README.md

# modelplotr: Plots to evaluate the business value of predictive models

The modelplotr package makes it easy to create a number of valuable
evaluation plots to assess the business value of a predictive model.
Using these plots, it can be shown how implementation of the model will
impact business targets like response or return on investment of a
campaign.

Plots available with modelplotr:

  - cumulative gains
  - cumulative lift
  - response
  - cumulative response
  - costs & revenues
  - profit
  - return on investment

Some benefits of using modelplotr:

  - *easy to explain* plots to discuss your model with business
  - *easy to use* on top of predictive models built with caret, mlr,
    h2o, keras or otherwise (with or without r)
  - supports both *binary and multinomial targets*
  - provides *four plotting scopes*:
      - comparing models
      - comparing datasets
      - comparing multiclass target classes
      - no comparison (single line)
  - *plot annotation*: highlighting specific values and adding
    explanatory text to guide interpretation
  - *plot customization*: all textual elements, line colors
  - *save plot* to file on disk

## Installation

You can install `modelplotr` from
[GitHub](https://github.com/modelplot/modelplotr) with:

``` r
devtools::install_github("modelplot/modelplotr")
```

See this blog for further details and examples of using the package.


NEWS.md
NEWS-modelplotr
================

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

    #> Last Update: 2019-04-13 13:17:21

## v1.0.0

2019-04-13 \* Initial CRAN release.


tests/testthat.R
library(testthat)
library(modelplotr)

test_check("modelplotr")


tests/testthat/test-prepare_scores_and_deciles.R
context("prepare_scores_and_deciles")

library(magrittr)
library(testthat)
# Generate some sample data -----------------------------------------------

data(iris)

# add some noise to iris to prevent perfect models
addNoise <- function(x) round(rnorm(n=100,mean=mean(x),sd=sd(x)),1)

iris_addnoise <- as.data.frame(lapply(iris[1:4], addNoise))

iris_addnoise$Species <- sample(unique(iris$Species),100,replace=TRUE)

iris <- rbind(iris,iris_addnoise)

train_index =  sample(seq(1, nrow(iris)),size = 0.7*nrow(iris), replace = F )
train = iris[train_index,]

test = iris[-train_index,]


# Expected answers --------------------------------------------------------

expected_cols <- c("model_label", "dataset_label", "y_true",
                   "prob_setosa", "prob_versicolor", "prob_virginica",
                   "dcl_setosa", "dcl_versicolor", "dcl_virginica")

expected_col_types <- c("factor", "factor", "factor", "numeric",
                        "numeric", "numeric", "numeric", "numeric",
                        "numeric")




# H2o tests ---------------------------------------------------------------

# Load H2o and initialise -----------------------------------------------
library(h2o)

h2o.init()
h2o.no_progress()

h2o_train <- train %>%
  as.h2o()

# Train a GBM -------------------------------------------------------------

h2o_model <- h2o.gbm(y = "Species",
                     x = setdiff(colnames(train), "Species"),
                     training_frame = h2o_train,
                     nfolds = 5)


df <- prepare_scores_and_deciles(datasets=list("train","test"),
                                 dataset_labels = list("train data","test data"),
                                 models = list("h2o_model"),
                                 model_labels = list("h2o gbm"),
                                 target_column="Species")

test_that("h2o models are properly formatted for aggregate_over_deciles", {
  df <- get("scores_and_deciles")
  col_names <- colnames(df)
  col_types <- map_chr(df, ~ class(.))

  expect_equal(colnames(df),  expected_cols)
  expect_equal(unname(col_types), expected_col_types)
})


# Test MLR support
trainTask <- mlr::makeClassifTask(data = train, target = "Species")
testTask <- mlr::makeClassifTask(data = test, target = "Species")

# Test mlr support --------------------------------------------------------

mlr::configureMlr() # this line is needed when using mlr without loading it (mlr::)

task <- mlr::makeClassifTask(data = train, target = "Species")
lrn <- mlr::makeLearner("classif.randomForest", predict.type = "prob")
rf <- mlr::train(lrn, task)


prepare_scores_and_deciles(datasets=list("train","test"),
                           dataset_labels = list("train data","test data"),
                           models = list("rf"),
                           model_labels = list("random forest"),
                           target_column="Species")

test_that("mlr models are properly formatted for aggregate_over_deciles", {

  df <- get("scores_and_deciles")
  col_names <- colnames(df)
  col_types <- map_chr(df, ~ class(.))

  expect_equal(colnames(df),  expected_cols)
  expect_equal(unname(col_types), expected_col_types)
})



# Test caret support ------------------------------------------------------

rf <- caret::train(Species ~.,data = train, method = "rf")

prepare_scores_and_deciles(datasets=list("train","test"),
                           dataset_labels = list("train data","test data"),
                           models = list("rf"),
                           model_labels = list("random forest"),
                           target_column="Species")

test_that("caret models are properly formatted for aggregate_over_deciles", {

  df <- get("scores_and_deciles")
  col_names <- colnames(df)
  col_types <- map_chr(df, ~ class(.))

  expect_equal(colnames(df),  expected_cols)
  expect_equal(unname(col_types), expected_col_types)
})



R/bank_td.R
#' Bank clients that have/have not subscribed a term deposit.
#'
#' A dataset containing some customer characteristics for clients of a bank that have/have not subscribed a term deposit.
#'
#' @format A data frame with 41118 rows and 6 variables:
#' \describe{
#'   \item{has_td}{has the client subscribed a term deposit? Values: "term deposit", "no".
#'   This variable is used as the binary target variable in examples for the modelplotr package.}
#'   \item{td_type}{what type of term deposit did the client subscribe? Values: "no.td", "td.type.A", "td.type.B", "td.type.C".
#'   This variable is used as the multinomial target variable in examples for the modelplotr package.}
#'   \item{duration}{last contact duration, in seconds (numeric)}
#'   \item{campaign}{number of contacts performed during this campaign and for this client}
#'   \item{pdays}{number of days that passed by after the client was last contacted from a previous campaign}
#'   \item{previous}{number of contacts performed before this campaign and for this client (numeric)}
#'   \item{euribor3m}{euribor 3 month rate}
#' }
#' @source This dataset is a subset of the dataset made available by the University of California, Irvine.
#' The complete dataset is available here: \url{https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank-additional.zip}
"bank_td"

# # SCRIPT TO CREATE bank_td
# library(dplyr)
# #zipname = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank-additional.zip'
# # we encountered that the source at uci.edu is not always available, therefore we made a copy to our repos.
# zipname = 'https://modelplot.github.io/img/bank-additional.zip'
# csvname = 'bank-additional/bank-additional-full.csv'
# temp <- tempfile()
# download.file(zipname,temp, mode="wb")
# bank <- read.table(unzip(temp,csvname),sep=";", stringsAsFactors=FALSE,header = TRUE)
# unlink(temp)
#
# summary(bank)
# bank <- bank[,c('y','duration','campaign','pdays','previous','euribor3m')]
#
# # definition of y_cat, created to show modelplotr for multinomial targets
# bank <- bank %>%
#   mutate(has_td = case_when(y =='yes' ~ 'term.deposit',
#                             y =='no' ~ 'no.term.deposit'),
#          has_td = factor(has_td,levels=c('term.deposit','no.term.deposit')),
#          td_value = rexp(nrow(bank),1/5000)+ bank$duration * 20 +
#            500*(bank$campaign*rnorm(nrow(bank),1,1))+bank$euribor3m*(100+rnorm(nrow(bank),200,100)),
#          td_type = case_when(y == 'no' ~ 'no.td',
#                            td_value < 10000 ~ 'td.type.A',
#                            td_value <= 25000 ~ 'td.type.B',
#                            TRUE ~ 'td.type.C'),
#          td_type = factor(td_type,levels = c('no.td','td.type.A','td.type.B','td.type.C'))) %>%
#   select(has_td,td_type,duration,campaign,pdays,previous,euribor3m)
#
# bank_td <- bank
# devtools::use_data(bank_td,overwrite = TRUE)



R/modelplotr.R
.onAttach <- function(libname, pkgname) {
  packageStartupMessage("Package modelplotr loaded! Happy model plotting!")
}


#' modelplotr: Plots to Evaluate the Business Performance of Predictive Models.
#'
#' Plots to evaluate the business performance of predictive models in R.
#' A number of widely used plots to assess the quality of a predictive model from a business perspective
#' can easily be created. Using these plots, it can be shown how implementation of the model will impact
#' business targets like response on a campaign or return on investment. It's very easy to apply modelplotr
#' to predictive models that are developed in caret, mlr, h2o or keras. For other models, even those built
#' outside of R, an instruction is included.
#' The modelplotr package provides three categories of important functions:
#' datapreparation, plot parameterization and plotting.
#'
#' @author Jurriaan Nagelkerke <jurriaan.nagelkerke@@gmail.com> [aut, cre]
#' @author Pieter Marcus <pieter.marcus@@persgroep.net> [aut]
#'
#' @section Datapreparation functions:
#'  The datapreparation functions are:
#' \describe{
#'   \item{\code{\link{prepare_scores_and_ntiles}}}{Function that builds a dataframe
#'   that contains actuals and predictions on the target variable for each dataset in \code{datasets} and each model in \code{models}.
#'   As inputs, it takes dataframes to score and model objects created with \strong{caret}, \strong{mlr}, \strong{h2o} or \strong{keras}.
#'   To use modelplotr on top of models created otherwise, even models built outside r, see \code{\link{aggregate_over_ntiles}}}
#'   \item{\code{\link{plotting_scope}}}{Function that creates a dataframe in the required format for all
#'   modelplotr plots, relevant to the selected scope of evaluation. Each record in this dataframe represents
#'   a unique combination of datasets, models, target classes and ntiles. As an input, plotting_scope can handle
#'   both a dataframe created with \code{aggregate_over_ntiles} as well as a dataframe created with
#'   \code{prepare_scores_and_ntiles} (or created otherwise, with similar layout). }
#'   \item{\code{\link{aggregate_over_ntiles}}}{Function that aggregates the output of \code{prepare_scores_and_ntiles}
#'   to create a dataframe with aggregated actuals and predictions. Each record in this dataframe represents
#'   a unique combination of datasets, models, target classes and ntiles. In most cases, you do not need to use function
#'   since the \code{plotting_scope} function will call this function automatically. }}
#' @section Parameterization functions:
#'  Most parameterization functions are internal functions. However, one is available for customization:
#' \describe{
#'   \item{\code{\link{customize_plot_text}}}{Function that returns a list that contains all textual elements for
#'   all plots that modelplotr can create. By changing the elements in this list - simply by overwriting values -
#'   and then including this list with the \code{custom_plot_text} parameter in plot functions, plot texts can easily be customized
#'   to meet your (language) preferences}}
#' @section Plotting functions:
#'   The plotting functions are:
#' \describe{
#'   \item{\code{\link{plot_cumgains}}}{Generates the cumulative gains plot. This plot, often referred to as the gains chart,
#'     helps answering the question: \strong{\emph{When we apply the model and select the best X ntiles,
#'     what percentage of the actual target class observations can we expect to target?}} }
#'     \item{\code{\link{plot_cumlift}}}{Generates the cumulative lift plot, often referred to as lift plot or index plot,
#'     helps you answer the question: \strong{\emph{When we apply the model and select the best X ntiles,
#'     how many times better is that than using no model at all?}}}
#'     \item{\code{\link{plot_response}}}{Generates the response plot. It plots the percentage of target class observations
#'     per ntile. It can be used to answer the following business question: \strong{\emph{When we apply
#'     the model and select ntile X, what is the expected percentage of target class observations
#'     in that ntile?}}}
#'     \item{\code{\link{plot_cumresponse}}}{Generates the cumulative response plot. It plots the cumulative percentage of
#'      target class observations up until that ntile. It helps answering the question:
#'      \strong{\emph{When we apply the model and select up until ntile X, what is the expected percentage of
#'      target class observations in the selection? }}}
#'     \item{\code{\link{plot_multiplot}}}{Generates a canvas with all four evaluation plots - cumulative gains, cumulative lift,
#'     response and cumulative response - combined on one canvas}
#'     \item{\code{\link{plot_costsrevs}}}{It plots the cumulative costs and revenues up until that ntile when the model
#'     is used for campaign selection. It can be used to answer the following business question:
#'      \strong{\emph{When we apply the model and select up until ntile X, what are the expected costs and
#'      revenues of the campaign?}}}
#'     \item{\code{\link{plot_profit}}}{Generates the Profit plot. It plots the cumulative profit up until that ntile when the
#'     model is used for campaign selection. It can be used to answer the following business question:
#'      \strong{\emph{When we apply the model and select up until ntile X, what is the expected profit of the campaign?}}}
#'     \item{\code{\link{plot_roi}}}{Generates the Return on Investment plot. It plots the cumulative revenues as a percentage
#'     of investments up until that ntile when the model is used for campaign selection. It can be used to answer the following
#'     business question: \strong{\emph{When we apply the model and select up until ntile X, what is the expected % return on
#'     investment of the campaign?}}}
#'     }
#'
#' @seealso \code{vignette('modelplotr')}
#' @seealso \url{https://github.com/modelplot/modelplotr} for details on the package
#' @seealso \url{https://modelplot.github.io/} for our blog posts on using modelplotr
#' @examples
#'\dontrun{
#' # load example data (Bank clients with/without a term deposit - see ?bank_td for details)
#' data("bank_td")
#'
#' # prepare data for training model for binomial target has_td and train models
#' train_index =  sample(seq(1, nrow(bank_td)),size = 0.5*nrow(bank_td) ,replace = FALSE)
#' train = bank_td[train_index,c('has_td','duration','campaign','pdays','previous','euribor3m')]
#' test = bank_td[-train_index,c('has_td','duration','campaign','pdays','previous','euribor3m')]
#'
#' #train models using caret... (or use mlr or H2o or keras ... see ?prepare_scores_and_ntiles)
#' # setting caret cross validation, here tuned for speed (not accuracy!)
#' fitControl <- caret::trainControl(method = "cv",number = 2,classProbs=TRUE)
#' # random forest using ranger package, here tuned for speed (not accuracy!)
#' rf = caret::train(has_td ~.,data = train, method = "ranger",trControl = fitControl,
#'                   tuneGrid = expand.grid(.mtry = 2,.splitrule = "gini",.min.node.size=10))
#' # mnl model using glmnet package
#' mnl = caret::train(has_td ~.,data = train, method = "glmnet",trControl = fitControl)
#'
#' # load modelplotr
#' library(modelplotr)
#'
#' # transform datasets and model objects to input for modelplotr
#' scores_and_ntiles <- prepare_scores_and_ntiles(datasets=list("train","test"),
#'                          dataset_labels = list("train data","test data"),
#'                          models = list("rf","mnl"),
#'                          model_labels = list("random forest","multinomial logit"),
#'                          target_column="has_td",
#'                          ntiles=100)
#'
#' # set scope for analysis (default: no comparison)
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles)
#' head(plot_input)
#'
#' # ALL PLOTS, with defaults
#' plot_cumgains(data=plot_input)
#' plot_cumlift(data=plot_input)
#' plot_response(data=plot_input)
#' plot_cumresponse(data=plot_input)
#' plot_multiplot(data=plot_input)
#' # financial plots - these need some financial parameters
#' plot_costsrevs(data=plot_input,fixed_costs=1500,variable_costs_per_unit=10,profit_per_unit=50)
#' plot_profit(data=plot_input,fixed_costs=1500,variable_costs_per_unit=10,profit_per_unit=50)
#' plot_roi(data=plot_input,fixed_costs=1500,variable_costs_per_unit=10,profit_per_unit=50)
#'
#' # CHANGING THE SCOPE OF ANALYSIS
#' # changing the scope - compare models:
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles,scope="compare_models")
#' plot_cumgains(data=plot_input)
#' # changing the scope - compare datasets:
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles,scope="compare_datasets")
#' plot_roi(data = plot_input,fixed_costs=1500,variable_costs_per_unit=10,profit_per_unit=50)
#' # changing the scope - compare target classes:
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles,scope="compare_targetclasses")
#' plot_response(data=plot_input)
# '
#' # HIGHLIGHTING OPTIONS
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles,
#'                              scope = 'compare_datasets',select_model_label = 'random forest')
#' plot_cumgains(data=plot_input,highlight_ntile=20)
#' plot_cumlift(data=plot_input,highlight_ntile=20,highlight_how = 'plot')
#' plot_response(data=plot_input,highlight_ntile=20,highlight_how = 'text')
#' plot_cumresponse(data=plot_input,highlight_ntile=20,highlight_how = 'plot_text')
#' plot_costsrevs(data=plot_input,fixed_costs = 1500,variable_costs_per_unit = 10,
#'                profit_per_unit = 50,highlight_ntile='max_roi')
#' plot_profit(data=plot_input,fixed_costs = 1500,variable_costs_per_unit = 10,profit_per_unit = 50)
#' plot_roi(data=plot_input,fixed_costs = 1500,variable_costs_per_unit = 10,profit_per_unit = 50)
#'
#' # OTHER PLOT CUSTOMIZATIONS
#' # customize line colors
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles,scope = 'compare_models')
#' plot_cumgains(data=plot_input,custom_line_colors = c('pink','navyblue'))
#' # customize all textual elements of plots
#' plot_input <- plotting_scope(prepared_input = scores_and_ntiles)
#' mytexts <- customize_plot_text(plot_input = plot_input)
#' mytexts$cumresponse$plottitle <- 'Expected conversion rate for Campaign XYZ'
#' mytexts$cumresponse$plotsubtitle <- 'proposed selection: best 15 percentiles according to our model'
#' mytexts$cumresponse$y_axis_label <- '% Conversion'
#' mytexts$cumresponse$x_axis_label <- 'percentiles (percentile = 1% of customers)'
#' mytexts$cumresponse$annotationtext <-
#' "Selecting up until the &NTL percentile with model &MDL has an expected conversion rate of &VALUE"
#' plot_cumresponse(data=plot_input,custom_plot_text = mytexts,highlight_ntile = 15)
#' }
#' @docType package
#' @name modelplotr
NULL


package_dev.R
###################################################################################
# GENERAL CHECKS
###################################################################################
# install.packages('devtools')
# install.packages('roxygen2')
# install.packages('testthat')
# install.packages('ggplot2')
# install.packages('gridExtra')
# install.packages('magrittr')
# install.packages('tidyverse')
# install.packages('RColorBrewer')
# install.packages('mlr')
# install.packages('randomForest')
# install.packages('knitr')
# install.packages('rmarkdown')
# install.packages("ggfittext")
??ggfittext
devtools::install_github("r-lib/devtools")
library(devtools)
library(roxygen2)

library(grid)

R.home()
file.edit(file.path("~", ".Rprofile")) # edit .Rprofile in HOME



RColorBrewer::display.brewer.all(n=4)
devtools::install_github('jurrr/modelplotr')
packageVersion("randomForest")
devtools::load_all('C:/TEMP/modelplotr')
library(roxygen2)
#library(modelplotr)

??modelplotr
?prepare_scores_and_deciles
?aggregate_over_deciles
?plotting_scope

?modelplotr

devtools::use_vignette("modelplotr")
devtools::use_testthat()
usethis::use_testthat()


# install package modelplotr from Github
devtools::install_github("jurrr/modelplotr")
library(modelplotr)
chooseCRANmirror()

chooseCRANmirror(graphics=FALSE, ind=1)
?modelplotr

?savemodelplots
savemodelplots("plot_cumgains")
###################################################################################
# PACHAGE EXAMPLE
###################################################################################




###################################################################################
# TEST WITH IRIS
###################################################################################

library(modelplotr)

# prepare iris dataset
data(iris)

# add some noise to iris to prevent perfect models
addNoise <- function(x) round(rnorm(n=100,mean=mean(x),sd=sd(x)),1)
iris_addnoise <- as.data.frame(lapply(iris[1:4], addNoise))
iris_addnoise$Species <- sample(unique(iris$Species),100,replace=TRUE)
iris <- rbind(iris,iris_addnoise)


# add some noise to iris to prevent perfect models

iris_addnoise = data.frame(
  Sepal.Length=round(rnorm(n=100,mean=mean(iris$Sepal.Length),sd=sd(iris$Sepal.Length)),1),
  Sepal.Width=round(rnorm(n=100,mean=mean(iris$Sepal.Width),sd=sd(iris$Sepal.Width)),1),
  Petal.Length=round(rnorm(n=100,mean=mean(iris$Petal.Length),sd=sd(iris$Petal.Length)),1),
  Petal.Width=round(rnorm(n=100,mean=mean(iris$Petal.Width),sd=sd(iris$Petal.Width)),1),
  Species=sample(unique(iris$Species),100,replace=TRUE))

iris <- rbind(iris,iris_addnoise)
test_size = 0.3
train_index =  sample(seq(1, nrow(iris)),size = (1 - test_size)*nrow(iris) ,replace = F )
train = iris[train_index,]
test = iris[-train_index,]
# estimate models with mlr
# this line is needed when usin mlr without loading it (mlr::)
mlr::configureMlr()
#estimate models
task = mlr::makeClassifTask(data = train, target = "Species")
lrn = mlr::makeLearner("classif.randomForest", predict.type = "prob")
rf = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.multinom", predict.type = "prob")
mnl = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.xgboost", predict.type = "prob")
xgb = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.lda", predict.type = "prob")
lda = mlr::train(lrn, task)


# apply modelplotr functions
prepare_scores_and_deciles(datasets=list("train","test"),
                      dataset_labels = list("train data","test data"),
                      models = list("rf","mnl","xgb","lda"),
                      model_labels = list("random forest","multinomial logit","XGBoost","Discriminant"),
                      target_column="Species")
#head(scores_and_deciles)
aggregate_over_deciles()
plotting_scope(scope = "compare_targetclasses")
plot_cumgains()
plot_cumgains(highlight_decile = 3)
plot_cumlift()
plot_cumlift(highlight_decile = 2)
plot_response()
plot_response(highlight_decile = 2)

aggregate_over_deciles(prepared_input = scores_and_deciles)
#head(deciles_aggregate)

plotting_scope(scope="compare_datasets",select_model_label = "random forest")
plotting_scope(scope="compare_models")
plotting_scope(scope="compare_targetclasses")
plotting_scope(scope="compare_datasets",select_model_label = "multinomial logit")
plotting_scope(scope = "compare_datasets",select_targetclass = NA)
plotting_scope(scope = "no_comparison",select_smallest_targetclass = FALSE)
plotting_scope(scope = "compare_models",select_smallest_targetclass = FALSE)
plotting_scope(scope = "compare_datasets",select_dataset_label = "train data")
plotting_scope(scope = "compare_targetclasses",select_dataset_label = list("train data","test data"))
plotting_scope(scope = "compare_targetclasses",select_targetclass = "setosa")
plotting_scope(scope = "compare_targetclasses", select_targetclass = list("setosa", "virginica"))
plotting_scope(scope = "compare_models",select_model_label = list("multinomial logit", "XGBoost"))
plotting_scope(scope = "compare_models",select_dataset_label = "train data" )
plotting_scope(scope = "compare_models", select_targetclass = list("setosa", "virginica"))
plotting_scope()
fourevalplots()

plot_cumgains()

plot_cumgains(highlight_decile = 2)

?modelplotr
?prepare_scores_and_deciles
?aggregate_over_deciles
?plotting_scope
#head(plot_input)

plot_cumgains(explainAtDecile = 3)

#`%>%` <- magrittr::`%>%`

#plot_input %>% dplyr::group_by(legend) %>% dplyr::summarize(n=n())
#test %>% dplyr::group_by(Species) %>% dplyr::summarize(n=n())

plot_cumgains1 <- plot_cumgains()
plot_cumgains1
plot_cumlift1 <- plot_cumlift()
plot_cumlift1
plot_response1 <- plot_response()
plot_response1
plot_cumresponse1 <- plot_cumresponse()
plot_cumresponse1
multiplot(plot_cumgains(),plot_cumlift(),plot_response(),plot_cumresponse(),cols=2)
str(fourevalplots())

# save plots
savemodelplots(c("plot_cumgains1","plot_cumlift1","plot_response1","plot_cumresponse1"))




###################################################################################
# test with bankingdata
###################################################################################

zipname = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank-additional.zip'
csvname = 'bank-additional/bank-additional.csv'

temp <- tempfile()
download.file(zipname,temp, mode="wb")
bank <- read.table(unzip(temp,csvname),sep=";", stringsAsFactors=FALSE,header = T)
bank <- bank[,c('duration','campaign','pdays','previous','euribor3m','y')]
summary(bank)
unlink(temp)

data = iris
task = mlr::makeClassifTask(data = bank, target = "y")
availableAlgorithms = mlr::listLearners(task, check.packages = FALSE)$name
print(paste0('Number of alorithms: ',length(availableAlgorithms)))
print(paste0('Algorithms (first 20): ',paste(availableAlgorithms[1:20],collapse = ', '),' ,...'))
?mlr::listLearners
bank$y = as.factor(bank$y)


test_size = 0.3
train_index =  sample(seq(1, nrow(bank)),size = (1 - test_size)*nrow(bank) ,replace = F )
train = bank[train_index,]
test = bank[-train_index,]

# subset with 4 columns
train <- train[,c('duration','campaign','pdays','previous','euribor3m','y')]
test <- test[,c('duration','campaign','pdays','previous','euribor3m','y')]

install.packages('xgboost')
# prepare data for training and train models
bank$y = as.factor(bank$y)
test_size = 0.3
train_index =  sample(seq(1, nrow(bank)),size = (1 - test_size)*nrow(bank) ,replace = F )
train = bank[train_index,]
test = bank[-train_index,]

# estimate models with mlr
library(mlr)
#models
task = makeClassifTask(data = train, target = "y")
lrn = makeLearner("classif.randomForest", predict.type = "prob")
rf = train(lrn, task)
lrn = makeLearner("classif.multinom", predict.type = "prob")
mnl = train(lrn, task)
lrn = makeLearner("classif.xgboost", predict.type = "prob")
xgb = train(lrn, task)
lrn = makeLearner("classif.lda", predict.type = "prob")
lda = train(lrn, task)


# estimate models with mlr
library(mlr)
#models
task = makeClassifTask(data = train, target = "y")
lrn = makeLearner("classif.randomForest", predict.type = "prob")
rf = train(lrn, task)
lrn = makeLearner("classif.multinom", predict.type = "prob")
mnl = train(lrn, task)
lrn = makeLearner("classif.xgboost", predict.type = "prob")
xgb = train(lrn, task)
lrn = makeLearner("classif.lda", predict.type = "prob")
lda = train(lrn, task)

# apply modelplotr
library(modelplotr)
prepare_scores_and_deciles(datasets=list("train","test"),
  dataset_labels = list("train data","test data"),
  models = list("rf","mnl","xgb","lda"),
  model_labels = list("random forest","multinomial logit","XGBoost","Discriminant"),
  target_column="y")
head(scores_and_deciles)

# estimate models with mlr
# this line is needed when usin mlr without loading it (mlr::)
mlr::configureMlr()
#models
task = mlr::makeClassifTask(data = train, target = "y")
lrn = mlr::makeLearner("classif.randomForest", predict.type = "prob")
rf = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.multinom", predict.type = "prob")
mnl = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.xgboost", predict.type = "prob")
xgb = mlr::train(lrn, task)
lrn = mlr::makeLearner("classif.lda", predict.type = "prob")
lda = mlr::train(lrn, task)

# apply modelplotr
prepare_scores_and_deciles(datasets=list("train","test"),
  dataset_labels = list("train data","test data"),
  models = list("rf","mnl","xgb","lda"),
  model_labels = list("random forest","multinomial logit","XGBoost","Discriminant"),
  target_column="y")
head(scores_and_deciles)
tail(scores_and_deciles)
dim(scores_and_deciles)

# apply modelplotr
library(modelplotr)
prepare_scores_and_deciles(datasets=list("train","test"),
  dataset_labels = list("train data","test data"),
  models = list("rf","mnl","xgb","lda"),
  model_labels = list("random forest","multinomial logit","XGBoost","Discriminant"),
  target_column="y")
head(scores_and_deciles)
aggregate_over_deciles()
plotting_scope()

aggregate_over_deciles()
tail(deciles_aggregate)

plotting_scope(scope="compare_datasets")
#plotting_scope(scope="compare_models")
#plotting_scope(scope="TargetValues",select_model_label="Discriminant")
head(plot_input)
tail(plot_input)

plotting_scope(scope="compare_datasets")
plotting_scope(scope="compare_models")
plotting_scope(scope="compare_targetclasses")
plotting_scope()

plotting_scope(select_model_label = "random forest",select_targetclass = "no",select_dataset_label = "train data")

plot_cumgains <- plot_cumgains(custom_line_colors=c("green"))
plot_cumgains + ggplot2::ggtitle("Gains plot") + ggplot2::theme(legend.position=c(0.975,0.025),legend.justification=c(1, 0))
plot_cumlift <- plot_cumlift()
plot_cumlift
plot_response <- plot_response()
plot_response
plot_cumresponse <- plot_cumresponse()
plot_cumresponse
multiplot(plot_cumgains,plot_cumlift,plot_response,plot_cumresponse,cols=2)

plot_cumresponse <- plot_cumresponse()
plot_cumresponse

# save plots
savemodelplots()

codetools::findGlobals(plot_response)

codetools::findGlobals(prepare_scores_and_deciles)
codetools::findGlobals(aggregate_over_deciles)
codetools::findGlobals(plotting_scope)

###################################################################################
# tutorial mdl
###################################################################################
# https://www.analyticsvidhya.com/blog/2016/08/practicing-machine-learning-techniques-in-r-with-mlr-package/
library(mlr)

summarizeColumns(train)



trainTask <- makeClassifTask(data = cd_train,target = "y")
testTask <- makeClassifTask(data = cd_test, target = "y")

getParamSet("classif.randomForest")

rf <- makeLearner("classif.randomForest", predict.type = "response", par.vals = list(ntree = 200, mtry = 3))
rf$par.vals <- list(importance = TRUE)

#set tunable parameters
#grid search to find hyperparameters
rf_param <- makeParamSet(
  makeIntegerParam("ntree",lower = 50, upper = 500),
  makeIntegerParam("mtry", lower = 3, upper = 4),
  makeIntegerParam("nodesize", lower = 10, upper = 50)
)

#let's do random search for 50 iterations
rancontrol <- makeTuneControlRandom(maxit = 50L)

#set 3 fold cross validation
set_cv <- makeResampleDesc("CV",iters = 3L)

#hypertuning
rf_tune <- tuneParams(learner = rf, resampling = set_cv, task = trainTask, par.set = rf_param, control = rancontrol, measures = acc)

#cv accuracy
rf_tune$y

#best parameters
rf_tune$x

#using hyperparameters for modeling
rf.tree <- setHyperPars(rf, par.vals = rf_tune$x)

#train a model
rforest <- train(rf.tree, trainTask)
getLearnerModel(rforest)

#make predictions
rfmodel <- predict(rforest, testTask)
predprob <- getPredictionProbabilities(rfmodel)
rfmodel$data

??mlr::predict


task = makeClassifTask(data = cd_train, target = "y")
getTaskDesc(task)$positive

lrn = makeLearner("classif.randomForest", predict.type = "prob")
rf = train(lrn, task)
# predict probabilities
pred = predict(rf, newdata = cd_test)

# Get probabilities for all classes
head(getPredictionProbabilities(pred))



??modelplotr

?plot_cumgains()
